modules/globebrowsing/cache/memoryawaretilecache.cpp

```diff
@@ -32,51 +35,204 @@ namespace globebrowsing {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    _singleton = new MemoryAwareTileCache(cacheSize);
+void MemoryAwareTileCache::create() {
+    _singleton = new MemoryAwareTileCache();
 }
 
 void MemoryAwareTileCache::destroy() {
-    std::lock_guard<std::mutex> guard(_mutexLock);
     delete _singleton;
 }
 
 MemoryAwareTileCache& MemoryAwareTileCache::ref() {
-    std::lock_guard<std::mutex> guard(_mutexLock);
     ghoul_assert(_singleton, "MemoryAwareTileCache not created");
     return *_singleton;
 }
 
 void MemoryAwareTileCache::clear() {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    _tileCache.clear();
+    _numTextureBytesAllocatedOnCPU = 0;
+    for (TextureContainerMap::iterator it = _textureContainerMap.begin();
+        it != _textureContainerMap.end();
+        it++)
+    {
+        it->second.first->reset();
+        it->second.second->clear();
+    }
+}
+
+void MemoryAwareTileCache::createDefaultTextureContainers() {
+    for (int id = 0; id < layergroupid::NUM_LAYER_GROUPS; id++) {
+        TileTextureInitData initData = LayerManager::getTileTextureInitData(layergroupid::ID(id));
+        assureTextureContainerExists(initData);
+    }
+}
+
+void MemoryAwareTileCache::assureTextureContainerExists(
+    const TileTextureInitData& initData)
+{
+    TileTextureInitData::HashKey initDataKey = initData.hashKey();
+    if (_textureContainerMap.find(initDataKey) == _textureContainerMap.end()) {
+        // For now create 500 textures of this type
+        _textureContainerMap.emplace(initDataKey,
+            TextureContainerTileCache(
+                std::make_unique<TextureContainer>(initData, 500),
+                std::make_unique<TileCache>(std::numeric_limits<std::size_t>::max())
+            )
+        );
+    }
+}
+
+void MemoryAwareTileCache::setSizeEstimated(size_t estimatedSize) {
+    ghoul_assert(_textureContainerMap.size() > 0, "Texture containers must exist.");
+    size_t sumTextureTypeSize = 0;
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
+        it != _textureContainerMap.cend();
+        it++)
+    {
+        sumTextureTypeSize +=
+            it->second.first->tileTextureInitData().totalNumBytes();
+    }
+    size_t numTexturesPerType = estimatedSize / sumTextureTypeSize;
+    resetTextureContainerSize(numTexturesPerType);
+}
+
+void MemoryAwareTileCache::resetTextureContainerSize(size_t numTexturesPerTextureType) {
+    _numTextureBytesAllocatedOnCPU = 0;
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
+        it != _textureContainerMap.cend();
+        it++)
+    {
+        it->second.first->reset(numTexturesPerTextureType);
+        it->second.second->clear();
+    }
 }
 
 bool MemoryAwareTileCache::exist(ProviderTileKey key) const {
-    std::lock_guard<std::mutex> guard(_mutexLock);
-    return _tileCache.exist(key);
+    for (TextureContainerMap::const_iterator it = _textureContainerMap.cbegin();
```

Comment:
I got it to work with
`std::find_if(_textureContainerMap.cbegin(), _textureContainerMap.cend(), [&](const std::pair<const TileTextureInitData::HashKey, TextureContainerTileCache>& p){ return p.second.second->exist(key); });`
Feels a bit strange since the `_textureContainerMap` is a
`std::unordered_map<TileTextureInitData::HashKey, TextureContainerTileCache>` (not `const` on the `HashKey`). Anyway it works. Not sure if I think it is much clearer than the for-loop. I guess I just have to get more used to the syntax.